iree/compiler/InputConversion/MHLO/MHLOToMHLOPreprocessing.cpp

```diff
@@ -674,6 +674,81 @@ class ScatterRank0Value : public OpRewritePattern<mhlo::ScatterOp> {
   }
 };
 
+// Mhlo of non-finite values (e.g. NaN, inf) and 0.0 produce 0.0 for XLA. For
+// linalg we need to conver these to select operations.
+class MulCastOfBool : public OpRewritePattern<mhlo::MulOp> {
+ public:
+  using OpRewritePattern<mhlo::MulOp>::OpRewritePattern;
+
+  // Traverse upward past common operations to see if the value came from a
+  // boolean tensor.
+  static bool IsFromBool(Value val) {
+    while (true) {
+      Operation *op = val.getDefiningOp();
+      if (!op) return false;
+
+      if (auto convertOp = dyn_cast<mhlo::ConvertOp>(op)) {
+        auto inTy = convertOp.operand().getType().cast<ShapedType>();
+        if (inTy.getElementType().isInteger(1)) {
+          return true;
+        }
+        val = convertOp.operand();
+        continue;
+      }
+
+      if (isa<mhlo::DynamicBroadcastInDimOp>(op) ||
+          isa<mhlo::BroadcastInDimOp>(op) || isa<mhlo::BroadcastOp>(op)) {
+        val = op->getOperand(0);
+        continue;
+      }
+
+      return false;
+    }
+  }
+
+  LogicalResult matchAndRewrite(mhlo::MulOp op,
+                                PatternRewriter &rewriter) const override {
+    auto resultTy = op.getType().cast<ShapedType>();
+    if (!resultTy.getElementType().isa<FloatType>()) return failure();
+    Value lhs = op.lhs();
+    Value rhs = op.rhs();
+    bool lhsIsBool = IsFromBool(lhs);
+    bool rhsIsBool = IsFromBool(rhs);
+    int64_t resultRank = resultTy.getRank();
+
+    if (lhsIsBool == rhsIsBool) return failure();
+    if (rhsIsBool) std::swap(lhs, rhs);
+
+    Type eType = resultTy.getElementType();
+    ShapedType lhsTy = lhs.getType().cast<ShapedType>();
+    Value lhsBool = rewriter.create<mhlo::ConvertOp>(
+        op.getLoc(), lhsTy.clone(rewriter.getIntegerType(1)), lhs);
+    Value zero = rewriter.create<mhlo::ConstOp>(
+        op.getLoc(), DenseElementsAttr::get(RankedTensorType::get({}, eType),
+                                            rewriter.getZeroAttr(eType)));
+
+    auto lhsShape = rewriter.create<shape::ShapeOfOp>(
+        op.getLoc(),
+        RankedTensorType::get({lhsTy.getRank()}, rewriter.getIndexType()), lhs);
+
+    auto broadcast = [&](Value value) -> Value {
+      auto valueTy = value.getType().cast<ShapedType>();
+      auto dimensions = llvm::to_vector<4>(
```

Comment:
Do we return `value` directly if broadcase is not needed? I.e., `value.getType() == resultTy`.